### ISSUE TITLE
Rethink the examples in examples/

### ISSUE THREAD

**Issue by masak** [masak/007#194]
I was really happy when we finally could close #54. Opened in October, closed it in March. (And finally got all the test in place in July.)

But I find that directory needs a higher bar nowadays. It should be about macros, (maybe) slangs, grammar hacking, language extensions, and the other things that makes 007 unique. It shouldn't just be examples that could be written in any half-competent ALGOLoid, no matter how nice they might look in 007.

Basically, I don't want people to go to the `examples/` directory and be underwhelmed.

Let's go through them. Anything that's just "ordinary code" takes the cut.

- [ ] 99-bottles.007: Remove.
- [ ] euclid-gcd.007: Remove.
- [ ] fizzbuzz.007: Remove.
- [x] format.007: This one has a macro. #183 will make it better. Keep.
- [ ] hanoi.007: Remove.
- [ ] hello-world.007: Remove.
- [x] name.007: Keep.
- [ ] nicomachus.007: Remove.
- [ ] quicksort.007: Remove.

Uh. But I also don't want just two examples in that directory.

So how about we also introduce new conforming examples with each one we remove? Yes, that'll make this issue decidedly non-low-hanging fruit. But I think it'll be worth it. It's a high ambition, but one we can comfortably aim for. There's plenty of macro ideas in the issue queue that we want to try anyway.

**Comment by masak** [comment]
I just want to add, for sentimentality, that once it *did* feel amazing that we had all those code examples. But 007 has basically outgrown that. The previously fantastic is now the mundane. And that's a good thing.

**Comment by masak** [comment]
I think two of the most realistic macro ideas to include in the examples would be #122 and #203. Both of those are definitely tied to real use cases.

**Comment by masak** [comment]
Oh, and if I were to dream a bit, #116 would be really nice to have in `examples/`. To the point where it would almost be worth *two* old example scripts. :smile:

But the thing is, 007 as interpreted through Rakudo is still too slow to make an effective demonstration of #116. So we'd need either a JavaScript backend (#166), or something similarly fast and impressive.

**Comment by masak** [comment]
I see I haven't mentioned #176 as an example I want in `examples/`. It definitely has a place there.

Yay! I love how this issue is now pushing the roadmap places. :smile:

**Comment by masak** [comment]
Heck, a `swap` macro! I just tried to write one, but...

    $ bin/007 -e='macro swap(a, b) { return quasi { my temp = {{{a}}}; {{{a}}} = {{{b}}}; {{{b}}} = temp; None } }; my w = 5; my z = 10; swap(w, z); say([w, z])'
    Missing quasi
    [...]

Dang, we really need to clean up our act with #10 and better parse errors in general. I want really good line/column numbers, and an arrow/⏏ into the place that's wrong.

As I tried to reduce this parse error, I ran into a simpler variant of #212. \*sigh\*

But anyway. `swap` macro. That takes our count to 6 macro examples. Since I said #116 is worth two old examples, we can actually throw out all the 7 old ones once we have those 6 new ones.

**Comment by masak** [comment]
I have two ideas for slangs. These should probably not be considered blockers for this issue, but I figure I'd jot them down.

* A Forth-like DSL inside a `forth { ... }` block term. The value of the whole term is what's on top of the stack at the end of the run. The powerful thing that makes this a slang is that values from the outside can be lexically accessed. Including custom operators. (Since notions of "infix" and "prefix" etc do not make sense inside the slang, some kind of pecking order has to be determined. What would we like `-` to mean, for example: `prefix:<->` or `infix:<->`? The boring solution would be to require the long names for everything; but that doesn't look so elegant.)

* An SQL DSL whose claim to fame would be that it doesn't allow injection attacks, because it escapes all the parametric values by default. This is an old idea; in 007 we don't really have to go all the way and connect to an actual database (although we could) &mdash; the main idea could be showcased somehow by just showing the resulting queries.

**Comment by masak** [comment]
Here, let me summarize the macro examples that we want to implement and replace the old examples with. I'll try to put them in order, most desirable first:

- [ ] `ff` and family (#207)
- [ ] `swap` macro (#218)
- [ ] reduce metaops (#176) ♁
- [ ] `postfix:<++>` and family (#122) ❦
- [ ] `+=` and family (#152) ❦♁
- [ ] `.=` mutating method call (#203) ❦♁
- [ ] `SEND + MORE = MONEY` (#116)

The ones marked ❦ need #214 to land in order to work non-buggily. The ones marked ♁ need `is parsed`.

**Comment by masak** [comment]
Also #202 and #215. My, there's no dearth of good macro examples to implement! (In my copious spare time, no less.) I would say importance-wise, these two slot in right above SEND MORE MONEY above.

**Comment by masak** [comment]
Also #158 since it nicely showcases nested macros.

**Comment by masak** [comment]
I think #229 (Elvis operator) deserves to go on this list as well.

**Comment by masak** [comment]
I know this issue isn't really accepting any more nominations... But as I'm thinking about #481 and what I'd like to have ready by then to talk about, I realize #186 would be a really nice, small example.

**Comment by masak** [comment]
Added a "currently blocked" label, because all the remaining examples are waiting on #177.
